File: packages/silo-client-java/src/test/java/in/org/quicko/silo/client/ScopeTest.java
package in.org.quicko.silo.client;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import in.org.quicko.silo.client.collections.CollectionSummary;
import in.org.quicko.silo.client.errors.NotFoundException;
import in.org.quicko.silo.client.scope.DeleteOptions;
import in.org.quicko.silo.client.scope.Project;
import in.org.quicko.silo.client.scope.ProjectHandle;
import in.org.quicko.silo.client.scope.RenameOptions;
import in.org.quicko.silo.client.scope.RenameReport;
import in.org.quicko.silo.client.support.StubHttp;
import in.org.quicko.silo.client.variables.Variable;
import java.time.Instant;
import java.util.List;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class ScopeTest {
  private StubHttp server;

  @BeforeEach
  void newStub() {
    server = StubHttp.create();
  }

  private Silo silo() {
    return server.silo("k");
  }

  @Test
  void aHandleMakesNoRequestUntilTheCallAtTheEndOfTheChain() {
    silo().project("acme").environment("prod").collection("posts");
    assertEquals(0, server.requestCount());
  }

  @Test
  void listsProjectsByIdAndName() {
    server.enqueueJson("{\"items\":[{\"id\":\"01P\",\"name\":\"acme\"}]}");
    List<Project> projects = silo().projects().list();

    assertEquals(List.of(new Project("01P", "acme")), projects);
  }

  @Test
  void readsARenameReportAndTheIdThatBindsTheRealCallToIt() {
    server.enqueueJson("{\"id\":\"01P\",\"from\":\"acme\",\"to\":\"acme-corp\","
        + "\"rewritten_claims\":[\"collections:acme/*/*:entries:read\"],"
        + "\"pattern_affected_claims\":[\"collections:acm*/*/*:entries:read\"]}");

    RenameReport report = silo().project("acme").rename("acme-corp", RenameOptions.preview());

    assertEquals("01P", report.id());
    assertEquals(List.of("collections:acme/*/*:entries:read"), report.rewrittenClaims());
    assertEquals(List.of("collections:acm*/*/*:entries:read"), report.patternAffectedClaims());
    assertTrue(server.lastRequest().target().contains("dry_run=true"));
  }

  @Test
  void aHandleKeepsAddressingTheNameItWasBuiltWith() {
    server.enqueueJson("{\"id\":\"01P\",\"from\":\"acme\",\"to\":\"acme-corp\","
        + "\"rewritten_claims\":[],\"pattern_affected_claims\":[]}");
    ProjectHandle handle = silo().project("acme");
    handle.rename("acme-corp");

    server.enqueue(404, "application/json",
        "{\"error\":{\"code\":\"not_found\",\"message\":\"no project named acme\"}}");

    assertThrows(NotFoundException.class, () -> handle.delete());
    assertEquals("/api/projects/acme", server.lastRequest().path());
  }

  @Test
  void sendsForceOnlyWhenItWasAskedFor() {
    server.enqueueNoContent();
    silo().project("acme").delete();
    assertEquals("/api/projects/acme", server.lastRequest().target());

    server.enqueueNoContent();
    silo().project("acme").delete(DeleteOptions.forced());
    assertEquals("/api/projects/acme?force=true", server.lastRequest().target());
  }

  @Test
  void mapsOnlyTheCollectionMetadataItKnows() {
    server.enqueueJson("{\"items\":[{\"id\":\"01C\",\"name\":\"posts\",\"entries\":42,"
        + "\"requires_auth\":true,\"created_at\":\"2026-09-01T10:00:00.000Z\","
        + "\"updated_at\":\"2026-09-02T10:00:00.000Z\"}]}");

    CollectionSummary summary = silo().scope("acme", "prod").collections().list().get(0);

    assertEquals(42, summary.entries());
    assertTrue(summary.requiresAuth());
    assertEquals(Instant.parse("2026-09-02T10:00:00Z"), summary.updatedAt());
  }

  @Test
  void tellsAnUnsetVariableValueFromAnEmptyOne() {
    server.enqueueJson("{\"items\":["
        + "{\"name\":\"API_URL\",\"description\":\"\",\"value\":null,\"set_in\":0,"
        + "\"created_at\":\"2026-09-01T10:00:00.000Z\",\"updated_at\":\"2026-09-01T10:00:00.000Z\"},"
        + "{\"name\":\"BANNER\",\"description\":\"\",\"value\":\"\",\"set_in\":1,"
        + "\"created_at\":\"2026-09-01T10:00:00.000Z\",\"updated_at\":\"2026-09-01T10:00:00.000Z\"}]}");

    List<Variable> variables = silo().scope("acme", "prod").variables().list();

    assertTrue(variables.get(0).value().isEmpty(), "an unset value leaves the reference standing");
    assertEquals("", variables.get(1).value().orElseThrow(), "an empty value is one somebody chose");
  }
}

File: packages/silo-client-dart/test/scope_test.dart
import 'package:silo_client/silo_client.dart';
import 'package:test/test.dart';

import 'support/stub_http.dart';

void main() {
  late StubHttp stub;
  late Silo silo;
  setUp(() {
    stub = StubHttp();
    silo = stub.silo();
  });

  test('a rename preview binds the real call to the record it described', () async {
    final report = {
      'id': '01P',
      'from': 'acme',
      'to': 'acme-corp',
      'rewritten_claims': ['entries:read:acme/*'],
      'pattern_affected_claims': <String>[],
    };
    stub
      ..json(report)
      ..json(report);

    final preview = await silo.project('acme').rename('acme-corp', const RenameOptions(dryRun: true));
    expect(stub.last.url.query, 'dry_run=true');
    expect(stub.lastBody, {'name': 'acme-corp'});
    expect(preview.rewrittenClaims, ['entries:read:acme/*']);

    await silo.project('acme').rename('acme-corp', RenameOptions(expectedId: preview.id));
    expect(stub.last.url.query, 'expected_id=01P');
  });

  test('delete sends force only when asked', () async {
    stub
      ..empty()
      ..empty();
    await silo.scope('acme', 'staging').delete();
    expect(stub.last.url.query, isEmpty);
    await silo.scope('acme', 'staging').delete(const DeleteOptions(force: true));
    expect(stub.last.url.toString(), 'http://silo.test/api/projects/acme/envs/staging?force=true');
  });

  test('maps collection summaries and creates with the schema as given', () async {
    stub
      ..json({
        'items': [
          {
            'id': '01C',
            'name': 'posts',
            'entries': 3,
            'requires_auth': true,
            'created_at': '2026-09-01T10:00:00Z',
            'updated_at': '2026-09-02T10:00:00+05:30',
          },
        ],
      })
      ..json({
        'id': '01C',
        'name': 'posts',
        'schema': {'type': 'object'},
      });

    final summary = (await silo.scope('acme', 'prod').collections.list()).single;
    expect(summary.requiresAuth, isTrue);
    expect(summary.updatedAt, DateTime.utc(2026, 9, 2, 4, 30));

    final definition = await silo.scope('acme', 'prod').collections.create('posts', {'type': 'object'});
    expect(stub.lastBody, {
      'name': 'posts',
      'schema': {'type': 'object'},
    });
    expect(definition.schema, {'type': 'object'});
  });

  test('declares a variable per project and values it per environment', () async {
    final view = {
      'name': 'API_URL',
      'description': 'public API',
      'value': null,
      'set_in': 0,
      'created_at': '2026-09-01T10:00:00Z',
      'updated_at': '2026-09-01T10:00:00Z',
    };
    stub
      ..json(view)
      ..json(view);

    await silo
        .project('acme')
        .variables
        .declare(
          'API_URL',
          const DeclareVariableOptions(description: 'public API', environment: 'prod', value: 'https://api'),
        );
    expect(stub.last.url.query, 'env=prod');
    expect(stub.lastBody, {'name': 'API_URL', 'description': 'public API', 'value': 'https://api'});

    final unset = await silo.scope('acme', 'staging').variables.unset('API_URL');
    expect(stub.last.method, 'DELETE');
    expect(unset.value, isNull);
  });
}
